File: frontend/src/types/analysis.ts
/**
 * Analysis domain types — mirrored from docs/API_CONTRACT.md §4.3
 * (Stage 07: engine-scored `analyzed` rows; pre-Stage-07 `segmented` rows
 * still read back with null scores). Backend: `app/schemas/analysis.py`.
 * Same names, same optionality.
 */

/** Pipeline stage. POST returns `analyzed`; `segmented` is legacy; `failed` reserved (nothing writes it yet — the report renders a failure state if one ever reads back). */
export type AnalysisStatus = "segmented" | "analyzed" | "failed";

/** How one requirement was detected (segmentation evidence, not a score). */
export type SegmentationStrategy =
  "requirement_id" | "decimal" | "numbered" | "bullet" | "paragraph";

export interface SegmentationMeta {
  strategy: SegmentationStrategy;
  confidence: number;
  start_offset: number;
  end_offset: number;
  line_start: number;
  line_end: number;
}

export type RequirementSeverity = "low" | "medium" | "high" | "critical";

/** One detector finding — nested under its requirement (no parent id). */
export interface AnalysisIssue {
  id: string;
  detector_id: string;
  category: string;
  severity: RequirementSeverity;
  phrase: string;
  start_offset: number;
  end_offset: number;
  reason: string;
  recommendation: string;
  ai_explanation: string | null;
}

/** One scored requirement — `severity` is null when no issues fired. */
export interface SegmentedRequirement {
  id: string;
  position: number;
  identifier: string | null;
  section: string | null;
  text: string;
  score: number | null;
  severity: RequirementSeverity | null;
  issues_count: number;
  suggested_rewrite: string | null;
  suggestion_source: "rule" | "ai" | null;
  segmentation: SegmentationMeta;
  issues: AnalysisIssue[];
}

export type AnalysisBand = "low" | "moderate" | "high" | "very_high";

export interface ScoreDeduction {
  issue_id: string;
  severity: RequirementSeverity;
  points: number;
}

/** Transparent breakdown: base + per-issue deductions + severity counts. */
export interface ScoreBreakdown {
  base: number;
  deductions: ScoreDeduction[];
  counts: Record<RequirementSeverity, number>;
}

/** Supplementary dimensions (each 100 − mapped deductions, clamped). */
export interface HealthDimensions {
  clarity: number;
  specificity: number;
  measurability: number;
  completeness: number;
}

/**
 * Analysis detail (POST result + GET by id). Issues live nested per
 * requirement; there is no top-level `issues` and no `source_excerpt`
 * (summary-only) in the detail.
 */
/** Source-document display pointer (`document` analyses only, else null). */
export interface AnalysisDocumentRef {
  filename: string;
  file_type: "pdf" | "docx" | "txt";
}

/** AI enhancement outcome (Stage 14): `ok` = overview (+ rewrites) generated;
 * `failed` = attempted, user-safe `ai_error` explains; `skipped` = not
 * requested; `unconfigured` = requested but no usable provider. */
export type AiStatus = "ok" | "failed" | "skipped" | "unconfigured";

/** POST /analysis/{id}/retry-ai — the four restamped AI fields only (Stage 17).
 * Clients re-read the full detail (fresh rewrites included) after a retry. */
export interface AiRetryResult {
  ai_status: AiStatus;
  ai_overview: string | null;
  ai_provider: string | null;
  ai_error: string | null;
}

export interface AnalysisResult {
  id: string;
  title: string;
  status: AnalysisStatus;
  source_type: "text" | "document";
  document: AnalysisDocumentRef | null;
  score: number | null;
  band: AnalysisBand | null;
  score_breakdown: ScoreBreakdown;
  requirements_count: number;
  issues_count: number;
  health: HealthDimensions | null;
  ai_overview: string | null;
  ai_provider: string | null;
  ai_status: AiStatus;
  ai_error: string | null;
  requirements: SegmentedRequirement[];
  created_at: string;
  updated_at: string;
}

/** History-list row: detail minus `requirements`, plus `source_excerpt`.
 * Stage 10: carries the `document` display pointer (like the detail). */
export interface AnalysisSummary {
  id: string;
  title: string;
  status: AnalysisStatus;
  source_type: "text" | "document";
  document: AnalysisDocumentRef | null;
  source_excerpt: string | null;
  score: number | null;
  band: AnalysisBand | null;
  requirements_count: number;
  issues_count: number;
  created_at: string;
  updated_at: string;
}

export type AnalysisSort = "created_at" | "-created_at" | "score" | "-score";

export interface ListAnalysesParams {
  page?: number;
  page_size?: number;
  sort?: AnalysisSort;
  band?: AnalysisBand;
  source_type?: "text" | "document";
  /** History search (Stage 10): title + document filename substring, ≤200 chars. */
  q?: string;
}

export interface CreateAnalysisInput {
  title: string;
  text: string;
  /** Opt into the post-commit AI step (Stage 14): overview + per-requirement
   * rewrites via the user's own provider; false = deterministic-only. */
  aiEnhance: boolean;
}

File: backend/app/schemas/analysis.py
"""Analysis API boundary models (API_CONTRACT §4.3 — Stage 08 subset).

TEXT posts here; FILE uploads go to POST /documents/upload (which reuses the
same pipeline and returns this same detail shape with `source_type` set
accordingly and `document` populated). `document_id` stays
accepted-but-rejected (400 — re-analyzing a stored document BY ID has no
pipeline yet) and `options.ai_enhance` stays
accepted-and-ignored until Stage 19. POST scores every requirement with the
deterministic engine (`analyzed`); pre-Stage-07 `segmented` rows still read
back with NULL scores.
"""

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

TITLE_MAX_LENGTH = 200  # mirrors analyses.title VARCHAR(200)
TEXT_MAX_LENGTH = 200_000  # contract §4.3: chars, enforced by the schema (400)


class AnalysisOptions(BaseModel):
    ai_enhance: bool = False


class AnalysisCreateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=TITLE_MAX_LENGTH)
    text: str = Field(min_length=1, max_length=TEXT_MAX_LENGTH)
    document_id: uuid.UUID | None = None
    options: AnalysisOptions = Field(default_factory=AnalysisOptions)

    @field_validator("title")
    @classmethod
    def _strip_title(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None  # blank → service fallback title

    @field_validator("text")
    @classmethod
    def _reject_blank_text(cls, value: str) -> str:
        # Whitespace-only input is a FIELD error (400 validation_error with a
        # `text` loc the UI maps to the editor) — not an attempted analysis.
        # The service still owns normalization; this only classifies blankness.
        if not value.strip():
            raise ValueError("text must not be blank")
        return value


class SegmentationMetaResponse(BaseModel):
    """How one requirement was segmented (evidence for the Stage 07 engine)."""

    strategy: str
    confidence: float
    start_offset: int
    end_offset: int
    line_start: int
    line_end: int


class IssueResponse(BaseModel):
    """One detector finding (contract §4.3 issue shape, exactly: no
    `requirement_id` — the nesting already says which requirement owns it)."""

    id: uuid.UUID
    detector_id: str
    category: str
    severity: str
    phrase: str
    start_offset: int
    end_offset: int
    reason: str
    recommendation: str
    ai_explanation: str | None = None


class RequirementResponse(BaseModel):
    """One scored requirement: `score` 0–100, `severity` = highest issue
    severity (None when clean), `suggested_rewrite` None until rewrite
    templates land (per-issue recommendations carry the guidance)."""

    id: uuid.UUID
    position: int
    identifier: str | None
    section: str | None = None
    text: str
    score: int | None = None
    severity: str | None = None
    issues_count: int = 0
    suggested_rewrite: str | None = None
    suggestion_source: str | None = None
    segmentation: SegmentationMetaResponse
    issues: list[IssueResponse] = Field(default_factory=list)


class DocumentRefResponse(BaseModel):
    """Source-document pointer for `source_type == "document"` (Stage 09):
    display metadata ONLY — filename + validated type. No document id (the
    UI needs no document link yet), no storage key, no binary, ever."""

    filename: str
    file_type: Literal["pdf", "docx", "txt"]


class AnalysisDetailResponse(BaseModel):
    """Analysis detail (contract §4.3, Stage 07 amendment: `status` is
    `analyzed` for engine-scored rows (`segmented` only for pre-Stage-07
    rows); `score`/`band`/`score_breakdown`/`health` populated by the
    engine; issues nested per requirement; no top-level `issues`, no
    `source_excerpt` (summary-only)."""

    id: uuid.UUID
    title: str
    status: Literal["segmented", "analyzed", "failed"]
    source_type: Literal["text", "document"]
    document: DocumentRefResponse | None = None
    score: int | None = None
    band: str | None = None
    score_breakdown: dict[str, Any] = Field(default_factory=dict)
    requirements_count: int
    issues_count: int
    health: dict[str, Any] | None = None
    ai_overview: str | None = None
    ai_provider: str | None = None
    ai_status: Literal["skipped"] = "skipped"
    ai_error: str | None = None
    requirements: list[RequirementResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class AnalysisSummaryResponse(BaseModel):
    """History-list row: detail minus `requirements`, plus `source_excerpt`.
    Stage 10: carries the `document` display pointer (like the detail)."""

    id: uuid.UUID
    title: str
    status: Literal["segmented", "analyzed", "failed"]
    source_type: Literal["text", "document"]
    document: DocumentRefResponse | None = None
    source_excerpt: str | None = None
    score: int | None = None
    band: str | None = None
    requirements_count: int
    issues_count: int
    created_at: datetime
    updated_at: datetime
